### DISCUSSION TITLE
fwupd offline update

### DISCUSSION THREAD

**Discussion by Aravindh09** [fwupd/fwupd#3643] [Q&A] [answered]
Hi @hughsie / @superm1,

I have couple of questions regarding offline update. can you please help me with this?

On ubuntu 20.04 based systems

Questions:
1) When bios password is set should we provide the password and unlock bios for fwupdmgr install --offline to go through?
2) if answer to question 1 is yes is there an option to provide the password from fwupdmgr itself so that bios can be unlocked before triggering fwupdmgr install --offline?

Thanks and regards,
Aravindh.

**Comment by superm1** [comment]
>When bios password is set should we provide the password and unlock bios for fwupdmgr install --offline to go through?

I can't imagine what the BIOS password is needed for this instance.  Can you elaborate?  `offline` is just a very minimal OS environment to install updates where there is a low liklihood of conflict from other applications and devices.  The only devices that need it are Super I/O plugin EC devices.  Others _can_ use it, but it would be overkill.

>if answer to question 1 is yes is there an option to provide the password from fwupdmgr itself so that bios can be unlocked before triggering fwupdmgr install --offline?

Can you describe the host device in further detail?  What is the relation to BIOS password to updating the device?  As I understand most common devices updated with fwupd don't have any needs for this.

**Reply by Aravindh09** [reply]
Hi @superm1,

Thank you for your previous response. 

The device base os is ubuntu 20.04 and has a bios admin password set so that the bios configs are not altered by the user. So any changes that need to be done related to bios configs would require a password which only the admin knows and the admin will have control over the device bios configs and upgrades.

So bios upgrade will also be administered by the admin. In this case a user who has access to terminal can initiate a bios upgrade using fwupdmgr with bios images from lvfs which the admin doesn't want the user to carry out. 

Basically the need is the user should not have control over the device's bios configs or bios upgrades and should only be administered by the admin.

The device could be part of an enterprise and connected to a central server managed by an admin.

Regards,
Aravindh.

**Reply by superm1** [reply]
>Basically the need is the user should not have control over the device's bios configs or bios upgrades and should only be administered by the admin.

Is the BIOS upgrade deployed using a UEFI capsule or something else?  UEFI capsules don't have a mechanism in the specification for this.  If this is not UEFI capsule, please describe further.  Is this Redfish?

>The device could be part of an enterprise and connected to a central server managed by an admin.

Is this a client or a server SUT?

**Reply by Aravindh09** [reply]
> 
> 
> > Basically the need is the user should not have control over the device's bios configs or bios upgrades and should only be administered by the admin.
> 
> Is the BIOS upgrade deployed using a UEFI capsule or something else? UEFI capsules don't have a mechanism in the specification for this. If this is not UEFI capsule, please describe further. Is this Redfish?
> 
Currently the bios upgrade is using the fwupd that comes bundled with ubuntu. I believe fwupd in turn uses the UEFI capsule update capabilities to perform the update. Please correct me if i m wrong.
> > The device could be part of an enterprise and connected to a central server managed by an admin.
> 
> Is this a client or a server SUT?

The device is a client device similar to Thin client devices managed by Wyse Management Suite.

**Reply by superm1** [reply]
>Currently the bios upgrade is using the fwupd that comes bundled with ubuntu. I believe fwupd in turn uses the UEFI capsule update capabilities to perform the update. Please correct me if i m wrong.

Correct.

>The device is a client device similar to Thin client devices managed by Wyse Management Suite.

So if it's focused around client like devices and UEFI capsule updates, last I checked - these don't need a BIOS password set to issue an update.  Once the policy is set in BIOS setup whether or not UEFI capsule updates are enabled, any OS can push down the capsule.
The BIOS will verify the signature of the payload (only Dell signed payloads are acceptable).

Has this changed?

**Reply by Aravindh09** [reply]
> 
> 
> > Currently the bios upgrade is using the fwupd that comes bundled with ubuntu. I believe fwupd in turn uses the UEFI capsule update capabilities to perform the update. Please correct me if i m wrong.
> 
> Correct.
> 
> > The device is a client device similar to Thin client devices managed by Wyse Management Suite.
> 
> So if it's focused around client like devices and UEFI capsule updates, last I checked - these don't need a BIOS password set to issue an update. Once the policy is set in BIOS setup whether or not UEFI capsule updates are enabled, any OS can push down the capsule.
> The BIOS will verify the signature of the payload (only Dell signed payloads are acceptable).
> 
> Has this changed?

Yes you are right. Only Dell signed payloads will be accepted by the active bios before an upgrade. The confusion or gap is to whether allow a user to initiate a bios upgrade even if he has a dell signed bios image. As per the clarification that you provided only if EDK supports bios upgrade after authentication using bios admin password, fwupd will be able to support it. Am i right?

Thanks a lot for your response.

Regards,
Aravindh.

**Comment by superm1** [comment, marked as answer]
>As per the clarification that you provided only if EDK supports bios upgrade after authentication using bios admin password, fwupd will be able to support it. Am i right?

The way that capsule update works is that `UpdateCapsule` UEFI runtime services is called to perform the update.  It just takes a buffer, and then OEM capsule handling code decides what to do after that.

BIOS admin password is an OEM concept, not part of EDK.  So any lockdown of what happens on the capsule itself would be OEM level customization. There are certainly ways you could do where you append some special binary content into the capsule that is passed down with some extra authentication, but I'll warn you I don't think this one is a good idea for two reasons:
1. You are going to **break** capsule updates on Windows 10/11.
2. You are going to need to change capsule runtime services handler from BIOS side, and fwupd side both.  Fwupd side needs to get re-signed, and possibly will break fwupd capsule handler ABI which has been stable for several years.  
3. You're going to need to have a way for the BIOS to advertise it requires this for fwupd to change behavior based upon.  Running either of them out of date is going to cause problems.

If you go down this path, we can review the PR to decide, but I strongly urge against it.

### Here's a different proposal for you:
Create a new Dell BIOS attribute let's say `CCMAuthentication`.  
1. This attribute will export from https://github.com/torvalds/linux/tree/master/drivers/platform/x86/dell/dell-wmi-sysman
2. This attribute when read will show whether extra authentication is needed to process capsules.  Let's say it offers 3 values: `required`, `not_required`, `authenticate`.
3. A extra plugin added to fwupd that will read this variable during uefi capsule updates on Dell systems if it's present.  If it's set to `not_required`, don't do anything.
4. If set to `required`, then try to read password off of a dbus interface published by the management software (only present when updates are allowed).  If dbus interface not available, then fail.
5. If set to `required` then use password to write `authenticate` into `CCMAuthentication` attribute
6. SMM handler for `CCMAuthentication` will cache the authorization for updates for the next boot somewhere in BIOS SPI.
7. UEFI capsule update is staged
8. System is restarted
3. Capsule flash routine verifies `CCMAuthentication` authorization and BIOS update signature.

Doing it this way you have a possibility to allow capsule updates to still work in Windows too.  Some Windows software can write CCMAuthentication attribute from WMI just like fwupd does.

**Reply by superm1** [reply]
One more idea- 
if you go this `CCMAuthentication` route you do have the possibility of having management software just do this all without fwupd's knowledge too if it's a libfwupd based client.  It can write the authentication stuff, and then call the update as `fwupdmgr` or any other libfwupd client would.

If someone tries to do it "by hand" with `fwupdmgr`/`gnome-software`/`kde-discover` would all "break" of course.  So the previous proposal is being a better open source citizen.

**Reply by Aravindh09** [reply]
Hi @superm1,

Thank you so much for your detailed response and inputs to my doubts. I am still new and exploring on various fwupd capabilities . Your suggestions and proposals are the key takeaways and learnings for me as part of this discussion.

Once again thanks a lot for your answers.

Will close this thread as I have got answers to my doubts.

Regards,
Aravindh.
